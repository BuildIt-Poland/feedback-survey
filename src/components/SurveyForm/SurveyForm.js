import React, { useRef } from 'react';
import { Redirect } from 'react-router-dom';
import styled from 'styled-components';
import { Box } from 'grommet';

import SurveyButton from '../Button';
import SurveyField from '../SurveyField';
import ScrollToFormikError from '../ScrollToError/ScrollToFormikError';
import { FINAL_PATH } from '../../constatnts/routes';
import StyledHeading from '../Layout/Heading';
import { colorWhite } from '../../styles/designTokens';
import Overlay from '../Overlay';
import { DURING_SUBMIT } from '../../constatnts/messages';

const ScrollAnchor = styled.div`
  position: absolute;
`;

const renderFormFields = (
  { questions, answerTypes },
  values,
  errors,
  touched,
  onChange,
  setFieldValue,
  scrollFieldRef
) =>
  questions.map((field, index) => {
    const isFirstError = Object.keys(errors)[0] === field.name;
    return (
      <React.Fragment key={index}>
        {isFirstError ? <ScrollAnchor ref={scrollFieldRef} /> : null}
        <SurveyField
          key={`field-${index}`}
          field={field}
          answerTypes={answerTypes}
          error={touched[field.name] ? errors[field.name] : null}
          value={values[field.name]}
          onChange={onChange}
          setFieldValue={setFieldValue}
        />
      </React.Fragment>
    );
  });

const SurveyForm = ({
  values,
  errors,
  touched,
  handleChange,
  handleSubmit,
  setFieldValue,
  isValid,
  submitCount,
  status,
  isSubmitting,
  surveyData
}) => {
  const scrollFieldRef = useRef(null);

  if (status && status.isSubmitted) {
    return (
      <Redirect
        to={{
          pathname: FINAL_PATH,
          state: { isSuccessSave: status.isSuccessSave, saveErrorMessage: status.saveErrorMessage }
        }}
      />
    );
  }

  if (isSubmitting) {
    return (
      <Overlay>
        <Box animation="pulse" align="center" justify="center" pad="medium" height="90vh">
          <StyledHeading level="2" color={colorWhite}>
            {DURING_SUBMIT}
          </StyledHeading>
        </Box>
      </Overlay>
    );
  }

  return (
    <>
      <ScrollToFormikError formik={{ isValid, submitCount, isSubmitting, errors }} scrollFieldRef={scrollFieldRef} />
      <form
        onSubmit={e => {
          e.preventDefault();
          handleSubmit();
        }}
      >
        {renderFormFields(surveyData, values, errors, touched, handleChange, setFieldValue, scrollFieldRef)}
        <Box align="center">
          <SurveyButton buttonType="submit" label="submit" margin="xlarge" data-test-id="survey-submit-button" />
        </Box>
      </form>
    </>
  );
};

export default SurveyForm;
