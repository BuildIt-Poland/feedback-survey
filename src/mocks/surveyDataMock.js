export const surveyDataMock = {
  surveyData: {
    questions: [
      {
        id: '4',
        type: '',
        content: 'Test question?',
        required: false,
        name: 'testName',
        answerType: 'kindOfFeedback'
      }
    ],
    answerTypes: [
      {
        type: 'kindOfFeedback',
        values: ['monthly', 'quaterly', 'annual']
      }
    ]
  },
  errors: {},
  values: {}
};
