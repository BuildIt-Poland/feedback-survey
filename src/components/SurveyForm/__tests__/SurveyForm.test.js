import React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent, cleanup } from 'react-testing-library';

import SurveyForm from '../SurveyForm';
import { surveyDataMock } from '../../../mocks/surveyDataMock';

jest.mock('react-router-dom', () => ({
  Redirect: 'Redirect'
}));

afterEach(cleanup);

describe('COMPONENT - SurveyForm', () => {
  it('renders correctly', () => {
    const component = create(<SurveyForm {...surveyDataMock} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with errors', () => {
    const updatedProps = {
      ...surveyDataMock,
      errors: {
        testName: 'error'
      }
    };
    const component = create(<SurveyForm {...updatedProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('calls handleSubmit when submit form', () => {
    const updatedProps = {
      ...surveyDataMock,
      handleSubmit: jest.fn()
    };
    const { getByText } = render(<SurveyForm {...updatedProps} />);

    fireEvent.click(getByText('submit'));

    expect(updatedProps.handleSubmit).toHaveBeenCalled();
  });

  it('shows saving survey placeholder when submit', () => {
    const updatedProps = {
      ...surveyDataMock,
      isSubmitting: true
    };
    const { getByText } = render(<SurveyForm {...updatedProps} />);

    expect(getByText('Saving survey...')).toBeDefined();
  });

  it('redirects after submit form', () => {
    const updatedProps = {
      ...surveyDataMock,
      status: {
        isSubmitted: true
      }
    };
    const component = create(<SurveyForm {...updatedProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
