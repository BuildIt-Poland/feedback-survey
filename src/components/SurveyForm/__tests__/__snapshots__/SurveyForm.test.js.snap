// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - SurveyForm redirects after submit form 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/final",
      "state": Object {
        "isSuccessSave": undefined,
        "saveErrorMessage": undefined,
      },
    }
  }
/>
`;

exports[`COMPONENT - SurveyForm renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="StyledGrommet-sc-19lkkz7-0 eUfLDp"
  >
    <div
      className="sc-bZQynM eCWoqb"
    >
      <div
        className="sc-EHOje dYFYCA FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy"
      >
        <label
          className="StyledText-sc-1sadyjn-0 bMXxti"
        >
          Test question?
        </label>
        <div
          className="StyledBox-sc-13pk1d4-0 gZSidW"
        >
          <div
            className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
          >
            <input
              autoComplete="off"
              className="StyledTextInput-sc-1x30a0s-0 cyWwXE"
              name="testName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="StyledBox-sc-13pk1d4-0 dNgQiZ"
  >
    <button
      className="sc-bdVaJa hZuWlY StyledButton-sc-323bzc-0 bmExQD"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      size="large"
      type="submit"
    >
      submit
    </button>
  </div>
</form>
`;

exports[`COMPONENT - SurveyForm renders correctly with errors 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="sc-dnqmqq kqTqqP"
  />
  <div
    className="StyledGrommet-sc-19lkkz7-0 eUfLDp"
  >
    <div
      className="sc-bZQynM eCWoqb"
    >
      <div
        className="sc-EHOje dYFYCA FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy"
      >
        <label
          className="StyledText-sc-1sadyjn-0 bMXxti"
        >
          Test question?
        </label>
        <div
          className="StyledBox-sc-13pk1d4-0 fVtecF"
        >
          <div
            className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
          >
            <input
              autoComplete="off"
              className="StyledTextInput-sc-1x30a0s-0 cyWwXE"
              name="testName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              value=""
            />
          </div>
        </div>
        <span
          className="StyledText-sc-1sadyjn-0 khlXyF"
        >
          error
        </span>
      </div>
    </div>
  </div>
  <div
    className="StyledBox-sc-13pk1d4-0 dNgQiZ"
  >
    <button
      className="sc-bdVaJa hZuWlY StyledButton-sc-323bzc-0 bmExQD"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      size="large"
      type="submit"
    >
      submit
    </button>
  </div>
</form>
`;
